Add intro to Web Animations blog post (#1431)
simplabs/simplabs.github.io: _posts/2021-01-29-web-animations-intro.md

keyframes which allows you to create more complicated animation effects which
are not possible with CSS transitions. CSS animations are usually triggered by
adding or removing a CSS class to an element with JavaScript.

```css
@keyframes fade-in-keyframes {
  0% {
    opacity: 0;
  }
  50% {
    opacity: 0.2;
  }
  100% {
    opacity: 1;
  }
}

.fade-in {
  animation: fade-in-keyframes 0.3s linear;
}
```

Both of these offer JavaScript events like `animationend` which allow you to
react to an animation that finished. They are a powerful, declarative and
performant way to do animations on individual elements.

### Animating with JavaScript

JavaScript based animation loops are the status quo for doing complex
animations. Animation libraries like
[ember-animated](https://ember-animation.github.io/ember-animated/) and
[framer-motion](https://www.framer.com/motion/) are built around this technique.

With a JavaScript based animation loop there is full control over timing and
state. This makes it possible to take other elements into account. For example,
you could take measurements of one element and animate another element from or

> to the position of the other.

Reviewer: when you say "other" are you referring to the first element? Id probably assume that but maybe changing it to "to the position of it" or "to the position of the first/initial element". let me know what you think, there are probably better ideas.

Author: I'll rewrite this part a bit to make it clearer.

Author: Rewritten it a bit to clarify. Would love another look at it!

```javascript
function move(transformEnd) {
  currentAnimation?.pause();
  let transformStart = getComputedStyle(element).transform;

  let progress = 1;
  if (currentAnimation) {
    progress = currentAnimation.effect.getComputedTiming().progress;
  }
  let duration = 1000 * progress;

  currentAnimation?.cancel();

  currentAnimation = element.animate(
    [{ transform: transformStart }, { transform: transformEnd }],
    {
      duration: duration,
      easing: 'linear',
      fill: 'forwards',
    },
  );
}
```

After this change our animations will always run with a constant velocity, no
matter when the running animation is cancelled.

TODO: insert example gif of the fully correct cancellation animation

### When an animation finishes

It is likely you need to do something after an animation completes. The Web
Animations API provides a couple of options to do this. Firstly there is the
`Animation.onfinish` handler which runs the given function after the animation
completes.

```javascript
currentAnimation.onfinish = () => console.log('animation finished!');
```

Secondly there is also the `Animation.finished` promise, which resolves when the
animation finishes.

```javascript

> currentAnimation.finished.then(() => console.log('animation finished!'));

Reviewer: Out of curiosity do these run when an animation is cancelled?

Author: They do not. In the case of the promise it throws an error which you need to catch, the onfinish hook is simply never called. In Ember land I wrap all this in ember-concurrency tasks which makes handling it quite elegant.

Author: Also, yes! This is the bit I needed to finish this section so thanks for asking.

<!--break-->

## The status quo of animating the web

Currently, there are two types of animation that are commonly use on the web.
CSS transitions/animations and animating through JavaScript by modifying inline
styles.

### CSS transitions and animations

[CSS transitions](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Transitions/Using_CSS_transitions)
are a basic but powerful way to define a transition between two CSS states of a
DOM element. They provide limited control, but due to their simplicity it's
often the preferred way to implement basic animations like hover effects on a
button. Transitions happen automatically between specified CSS variables with a
given timing function, delay and duration.

```css
.my-button {
  opacity: 0.5;
  transition: opacity 0.3s linear;
}

.my-button:hover {
  opacity: 1;
}
```

[CSS animations](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Animations/Using_CSS_animations)
offer more control over timing and easing. They can run for a given duration,
have a delay and an iteration count. The animation itself is specified through
keyframes which allows you to create more complicated animation effects which

> are not possible with CSS transitions. CSS animations are usually triggered by

Reviewer 2: It would be ideal to maybe break the sentence down in smaller ones, so 'which' is not repeating. I also believe it would be keyframes which **allow**.

Author: I've rewritten the sentence a bit, hope this is better.

adding or removing a CSS class to an element with JavaScript.

```css
@keyframes fade-in-keyframes {
  0% {
    opacity: 0;
  }
  50% {
    opacity: 0.2;
  }
  100% {
    opacity: 1;
  }
}

.fade-in {
  animation: fade-in-keyframes 0.3s linear;
}
```

> Both of these offer JavaScript events like `animationend` which allow you to

Reviewer 3: Would it be worth mentioning that although these events exist it's kind of hard to synchronize animations with application state?

Author: Yeah indeed. It's in there _very_ implicitly with the "you need to add/remove classes to trigger these animations". I'll expand a bit on it.

react to an animation that finished. They are a powerful, declarative and
performant way to do animations on individual elements.

### Animating with JavaScript

JavaScript based animation loops are the status quo for doing complex
animations. Animation libraries like
[ember-animated](https://ember-animation.github.io/ember-animated/) and

> [framer-motion](https://www.framer.com/motion/) are built around this technique.

Reviewer 3: This could be a nice opportunity to give a shoutout to tolls like script.aculo.us that did this about 10 years ago already.

to get an approximate 60 frames per second animation similar to the example
above. Nowadays the `requestAnimationFrame` API exists which results in smoother
animation as the browser will call it right before the next repaint. Browsers
can also pause `requestAnimationFrame` when a tab is out of focus to prevent
inactive tabs from running expensive animations at high framerates.

In addition, this animation technique requires the need for manual interpolation
of the values we might want to animate. This can be difficult for complex values
like color or clipPath, yet also gives opportunity to do different kinds of
interpolation than what the browser does by default, which can be especially
relevant for non-trivial values like a color space.

While this animation technique is very powerful it does result in a lot of
calculations being made every frame for the duration of the animation. This
might result in stutter or delay especially if something else is happening on
the page. Examples are a router transition to another page or some data fetching
and parsing. The browser also has limited ways to optimise such animations as it

> only knows about the current frame.

Reviewer 3: I assume these animations can also not be performend on the GPU (as opposed to CSS transitions/animations and presumable the animations API)? I guess saying something around this being kind of the brute force approach in lack for something better might help understand the situation better :)

Author: Yes indeed they cannot completely be performed on the GPU, which is also why it will stutter etc. when some heavy JS calculations are happening at the same time. CSS/Web Animations happen on the GPU, so they should not be impacted by JS as much.

## Web Animations API

The Web Animations API is a relatively new addition to the browser and is still
very much in development. It promises to combine the benefits of CSS

> Transitions/Animations and JavaScript based animations.

Reviewer 3: Might want to give a shoutout to @jayjayjpg here who has written a few blog posts and given (several I think) conference talks on the matter.

All popular browsers have now implemented the minimum features necessary to do
complex animations and a
[polyfill](https://github.com/web-animations/web-animations-js) which adds
support for the Web Animations API to older browsers is available. It's time we
tried it out!

### The basics

The main interface for Web Animations is the `Element.animate` function. It can
be called on a DOM Element, takes an array of keyframes as the first argument
and options as the second argument. On first sight the API is very similar to
CSS animations.

```html
<div id="circle"></div>

<script type="text/javascript">
  let element = document.getElementById('circle');
  element.animate([{ opacity: 0 }, { opacity: 1 }], {
    duration: 300,
    easing: 'linear',
  });
</script>
```

We'll need only a "from" and a "to" keyframe for now. The easing we set will
define the timing function used to interpolate between the keyframes. For now,
we will use the most basic one: "linear".

In addition to that there's some useful timing related functionality on the
animation instance we can use.

```javascript
// pause the animation

> animation.pause();

Reviewer 3: I assume `animation` is returned from `Element.prototype.animate`? It might be worth mentioning that to make that clear.

// timing information about the current state of the animation
let timing = animation.effect.getComputedTiming();

// progress of the animation between 0 and 1
timing.progress;

// resume the animation
animation.play();

// promise that will resolve when the animation has finished
animation.finished.then(() => { ... });
```

With these building blocks we can start creating more complex animations.

### Moving an element around

Let's create an animation which applies a CSS transform to move the element from
one position to another.

```javascript
element.animate(
  [{ transform: 'translateX(0)' }, { transform: 'translateX(100px)' }],
  {
    duration: 1000,
    easing: 'linear',
  },
);
```

This will transform the element 100 pixels to the right from its original
position in 1 second with a linear timing function. When the animation is done,

> it's effects on the element are automatically removed.

Reviewer 2: **its** effects:)

Reviewer 2: Overall: I unfortunately can't judge the content much but in terms of the writing style, it's very good to read.

Author: Thanks for the review!

TODO: insert example gif

Like with CSS animations we need to specify a fill option to specify in what way
we want the animation's effects to be retained. We can specify `fill: forwards`
so our final state is retained. Note that you will not see the animated styles

> appear in the DOM as you might be used to with JavaScript based animation.

Reviewer 3: I assume they'd show up in the styles inspector though?

Reviewer 3: I guess it could be cool to show how these animations integrate with the inspector anyway (e.g. are there debugging tools, can they be paused, inspected etc.) I don't think it'd be necessary to go into a lot of detail, just maybe a screenshot showing a bit of what's possible?

Author: There is a `commitStyles` feature which applies the state as inline CSS. The feature is unfortunately still very much in development https://github.com/w3c/csswg-drafts/issues/5394 and relying on that behaviour for now seems hard as it might change.

I indeed wanted to add a section about the inspector. They show up under the animations debugger in the same way as CSS transitions/animations do (you can also replay them, slow them down etc). As you also think this is a good addition, I'll add a small section.

Author: > I assume they'd show up in the styles inspector though?

Only as computed styles.

Author: A newer comment exists which talks about adding a section about the debugger, so I'll resolve this one.

---
title: 'An intro to animating with the Web Animations API'
author: 'Nick Schot'
github: nickschot
twitter: nickschot
topic: javascript
bio: 'Senior Software Engineer'
og:
  image: /assets/images/posts/2021-01-29-web-animations-intro/og-image.png
---

> Animations can be a useful tool to enhance user experience on the web. Aside

Reviewer 4: ```suggestion
Animations can be a useful tool to enhance the user experience on the web. Aside
```

The grammar feels off to me without it 😅

from providing a pleasant visual experience, animations can aid in the user's
understanding of elements appearing, moving and disappearing from a page. This
blog post will provide a short overview of the status quo of animating the web
and take an initial look at the current capabilities of the [Web Animations
API].

[web animations api]:
  https://developer.mozilla.org/en-US/docs/Web/API/Web_Animations_API

<!--break-->

![An intro to animating with the Web Animations API](/assets/images/posts/2021-01-29-web-animations-intro/illustration.svg#full)

## The status quo of animating the web

> Currently, there are two types of animation that are commonly use on the web.

Reviewer 4: ```suggestion
Currently, there are two types of animation that are commonly used on the web.
```

With a JavaScript based animation loop there is full control over timing and
state. This makes it possible to take other elements into account. For example,
you could take measurements of an element A and animate an element B from the
position and size of element A to element B's natural position.

A basic animation loop takes a source and target state and based on a timing
function calculate what the state for the current frame should be.

```javascript
let sourceOpacity = 0;
let targetOpacity = 1;
let duration = 500;
let time = 0;
// target 60 frames per second
let frameDuration = 1000 / 60;

let timer = setInterval(function () {
  // stop animation if time exceeds duration
  if (time > duration) {
    clearInterval(timer);
    return;
  }

  // calculate where we are in the animation as a factor between 0 and 1
  let factor = time / duration;

  // multiply the source/target diff with the calculated factor
  let opacity = sourceOpacity + (targetOpacity - sourceOpacity) * factor;

  // apply calculated opacity
  element.style.opacity = opacity;

  // move to the next frame
  time += frameDuration;
}, frameDuration);
```

> In the past animation loops with JavaScript had to be done through `setInterval`

Reviewer 4: ```suggestion
In the past, animation loops with JavaScript had to be done through `setInterval`
```

---
title: 'An intro to animating with the Web Animations API'
author: 'Nick Schot'
github: nickschot
twitter: nickschot
topic: javascript
bio: 'Senior Software Engineer'
og:
  image: /assets/images/posts/2021-01-29-web-animations-intro/og-image.png
---

Animations can be a useful tool to enhance the user experience on the web. Aside
from providing a pleasant visual experience, animations can aid in the user's
understanding of elements appearing, moving and disappearing from a page. This
blog post will provide a short overview of the status quo of animating the web
and take an initial look at the current capabilities of the [Web Animations
API].

[web animations api]:
  https://developer.mozilla.org/en-US/docs/Web/API/Web_Animations_API

<!--break-->

![An intro to animating with the Web Animations API](/assets/images/posts/2021-01-29-web-animations-intro/illustration.svg#full)

## The status quo of animating the web

> Currently, there are two types of animation that are commonly used on the web:

Author: "two animation techniques" might be better?

> Currently, there are two types of animation that are commonly used on the web:

Reviewer 3: indeed sounds better 👍

function calculate what the state for the current frame should be.

```javascript
let sourceOpacity = 0;
let targetOpacity = 1;
let duration = 500;
let time = 0;
// target 60 frames per second
let frameDuration = 1000 / 60;

let timer = setInterval(function () {
  // stop animation if time exceeds duration
  if (time > duration) {
    clearInterval(timer);
    return;
  }

  // calculate where we are in the animation as a factor between 0 and 1
  let factor = time / duration;

  // multiply the source/target diff with the calculated factor
  let opacity = sourceOpacity + (targetOpacity - sourceOpacity) * factor;

  // apply calculated opacity
  element.style.opacity = opacity;

  // move to the next frame
  time += frameDuration;
}, frameDuration);
```

In the past, animation loops with JavaScript had to be done through
`setInterval` to get an approximate 60 frames per second animation similar to
the example above. Nowadays the `requestAnimationFrame` API exists which results
in smoother animation as the browser will call it right before the next repaint.
Browsers can also pause `requestAnimationFrame` when a tab is out of focus to

> prevent inactive tabs from running expensive animations at high framerates.

Author: Yeah possibly, or both. They're really similar though. I could abstract the generic parts to a separate function and show both?

> prevent inactive tabs from running expensive animations at high framerates.

Reviewer 3: What's the benefit from showing the old way? Would one still have to use that in some scenarios? Otherwise I'd argue it's irrelevant?

> prevent inactive tabs from running expensive animations at high framerates.

Author: Purely (historical) context I guess. We can also replace it entirely.

> prevent inactive tabs from running expensive animations at high framerates.

Reviewer 3: I think I'd only show `requestAnimationFrame` in the code example then and only mention `setTimeout` in the text as the old but inferior approach for the same thing as you do already.

In addition, this animation technique requires the need for manual interpolation
of the values we might want to animate. This can be difficult for complex values
like color or clipPath, yet also gives opportunity to do different kinds of
interpolation than what the browser does by default, which can be especially
relevant for non-trivial values like a color space.

While this animation technique is very powerful it does result in a lot of
calculations being made every frame for the duration of the animation. This
might result in stutter or delay especially if something else is happening on
the page. Examples are a router transition to another page or some data fetching
and parsing. The browser also has limited ways to optimise such animations as it
only knows about the current frame.

## Web Animations API

> The Web Animations API is a relatively new addition to the browser and is still

Reviewer 3: I'd probably link to MDN again here. It's linked in the intro of course but I think it'd be convenient for readers to have it linked here again.

// promise that will resolve when the animation has finished
animation.finished.then(() => { ... });
```

With these building blocks we can start creating more complex animations.

### Moving an element around

Let's create an animation which applies a CSS transform to move the element from
one position to another.

```javascript
element.animate(
  [{ transform: 'translateX(0)' }, { transform: 'translateX(100px)' }],
  {
    duration: 1000,
    easing: 'linear',
  },
);
```

This will transform the element 100 pixels to the right from its original
position in 1 second with a linear timing function. When the animation is done,
its effects on the element are automatically removed.

![Basic move animation](/assets/images/posts/2021-01-29-web-animations-intro/video1.mp4#video)

Like with CSS animations we need to specify a fill option to specify in what way
we want the animation's effects to be retained. We can specify `fill: forwards`
so our final state is retained. Note that you will not see the animated styles
appear in the DOM as you might be used to with JavaScript based animation. You
can see the effect they have in the computed styles and animation sections of

> the browser's developer tools.

Reviewer 3: I think a screenshot of that would be cool.

Author: Yes indeed, this was one of the things I named this morning that I still wanted to add. I don't know exactly how yet as if I'd just add it here it'd be an image just above the GIF. I'll have to reorganise a bit or make it a separate section about debugging animations?

Reviewer 3: I guess a dedicated section or paragraph would be great!

Author: I decided to integrate this bit into the animation cancellation section to illustrate why it's breaking initially. Seems like the best place.

Alternatively the `animation.commitStyles()` feature can be used to put the
current animation state in the DOM as inline styles. The animation itself can
then safely be cancelled so the browser can free up the resources it consumes.

### Cancelling and resuming an animation

Cancellation is an important part of real-world animations. Animations take time
to complete and in that time the user might have interacted with the page in a
way where the running animation does not make sense anymore.

In theory, you only need to specify a "to" keyframe which would make it trivial
to cancel an animation as the browser will calculate the "from" position based
on the current state. Unfortunately not all browsers currently support that
behaviour meaning we'll have to be a little more elaborate for now.

In order to cancel the current animation (if any) we'll need to take a few
steps. Before trying to calculate our starting keyframe we'll need to pause the
currently running animation (if any). Then we'll need to get the current
transform present on our element. Finally, we can start the new animation.

Let's abstract our animation into a `move` function which takes a target

> `transform` string. We'll also add `animateRight` and `animateLeft` functions

Reviewer 3: It takes `transformStart` and `transformEnd` though?

Author: Oh good catch! This text is based on the "final solution" it seems. I'll fix it.

from providing an appealing visual experience, animations can aid in the user's
understanding of elements appearing, moving and disappearing from a page. This
blog post will provide a short overview of the status quo of animating the web
and take an initial look at the current capabilities of the [Web Animations
API].

[web animations api]:
  https://developer.mozilla.org/en-US/docs/Web/API/Web_Animations_API

<!--break-->

![An intro to animating with the Web Animations API](/assets/images/posts/2021-01-29-web-animations-intro/illustration.svg#full)

## The status quo of animating the web

Currently, there are two animation techniques that are commonly used on the web:
CSS transitions/animations and animating through JavaScript by modifying inline
styles.

### CSS transitions and animations

[CSS transitions](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Transitions/Using_CSS_transitions)
are a basic but powerful way to define a transition between two CSS states of a
DOM element. They provide limited control, but due to their simplicity it's
often the preferred way to implement basic animations like hover effects on a

> button. Transitions happen automatically between specified CSS variables with a

Reviewer 3: ```suggestion
button. Transitions happen automatically between values of a CSS attribute with a
```

?

individual elements.

### Animating with JavaScript

JavaScript based animation loops are the status quo for doing complex
animations. Animation libraries like
[ember-animated](https://ember-animation.github.io/ember-animated/) and
[framer-motion](https://www.framer.com/motion/) or older libraries like
[Velocity.js](http://velocityjs.org/) and
[script.aculo.us](https://github.com/madrobby/scriptaculous) are built around
this technique.

With a JavaScript based animation loop there is full control over timing and
state. This makes it possible to take other elements into account. For example,
you could take measurements of an element A and animate an element B from the
position and size of element A to element B's natural position.

A basic animation loop takes a source and target state and based on a timing
function calculate what the state for the current frame should be.

```javascript
let sourceOpacity = 0;
let targetOpacity = 1;
let duration = 500;

let startTime;
function step(time) {
  if (!startTime) {
    startTime = time;
  }

  let progress = time - startTime;

  // calculate where we are in the animation as a factor between 0 and 1

>   let factor = Math.max(progress / duration, 1);

Reviewer 3: I always find it hard to wrap my head around whether `max` or `min` is right in a given situation but should this not be

```suggestion
  let factor = Math.min(progress / duration, 1);
```

since otherwise it'd always be 1?

  // multiply the source/target diff with the calculated factor
  let currentOpacity = sourceOpacity + (targetOpacity - sourceOpacity) * factor;

  // apply calculated opacity
  element.style.opacity = currentOpacity;

  // move to the next frame if we haven't reached the target duration yet
  if (progress < duration) {
    window.requestAnimationFrame(step);
  }
}

// start the animation
window.requestAnimationFrame(step);
```

In the past, animation loops with JavaScript had to be done through
`setInterval` to get an approximate 60 frames per second animation. Nowadays the
`requestAnimationFrame` API exists which results in smoother animation as the
browser will call it right before the next repaint. Browsers can also pause
`requestAnimationFrame` when a tab is out of focus to prevent inactive tabs from
running expensive animations at high framerates.

In addition, this animation technique requires the need for manual interpolation
of the values we might want to animate. This can be difficult for complex values

> like color or clipPath, yet also gives opportunity to do different kinds of

Reviewer 3: ```suggestion
like `color` or `clipPath`, yet also gives opportunity to do different kinds of
```

so our final state is retained.

```javascript
element.animate(
  [{ transform: 'translateX(0)' }, { transform: 'translateX(100px)' }],
  {
    duration: 1000,
    easing: 'linear',
    fill: 'forwards',
  },
);
```

Our animation now behaves as we would expect and retains the styles specified in
the final keyframe.

![Basic move animation that retains final state](/assets/images/posts/2021-01-29-web-animations-intro/video2.mp4#video)

Alternatively the `animation.commitStyles()` feature can be used to put the
current animation state in the DOM as inline styles. The animation itself can
then safely be cancelled so the browser can free up the resources it consumes.

### Cancelling and resuming an animation

Cancellation is an important part of real-world animations. Animations take time
to complete and in that time the user might have interacted with the page in a
way where the running animation does not make sense anymore.

In theory, you only need to specify a "to" keyframe which would make it trivial
to cancel an animation as the browser will calculate the "from" position based
on the current state. Unfortunately not all browsers currently support that

> behaviour meaning we'll have to be a little more elaborate for now.

Reviewer 3: Does this not describe the situation where you'd want to cancel the current animation and immediately start **another** one with the element's current state? Before this you're talking about cancelling only.

Reviewer 3: I'd maybe just add something explicitly saying that usually when you cancel you want to replace the running animation with another animation or so to make it clear – don't feel this would require a major rephrasing. Just implying it (although I agree that's usually what you'd want probably) could be confusing.

currently running animation (if any). Then we'll need to get the current
transform present on our element. Finally, we can start the new animation.

Let's abstract our animation into a `move` function which takes `transformStart`
and `transformEnd` string arguments. We'll also add `animateRight` and
`animateLeft` functions which are hooked up to two buttons, so we can run the
animations by clicking those buttons.

```html
<button onclick="animateLeft()">Left</button>
<button onclick="animateRight()">Right</button>
<div id="circle"></div>

<script>
  const element = document.getElementById('circle');

  let currentAnimation;

  function move(transformStart, transformEnd) {
    currentAnimation = element.animate(
      [{ transform: transformStart }, { transform: transformEnd }],
      {
        duration: 1000,
        easing: 'linear',
        fill: 'forwards',
      },
    );
  }

  function animateRight() {
    move('translateX(0)', 'translateX(100px)');
  }

  function animateLeft() {
    move('translateX(100px)', 'translateX(0)');
  }
</script>
```

Rapidly clicking the buttons interchangeably now shows a pretty jarring
animation!

> ![Basic move animation](/assets/images/posts/2021-01-29-web-animations-intro/video3.mp4#video)

Reviewer 3: This doesn't render for me in Safari somehow (nor do any of the videos after this one) 🤔 

<img width="876" alt="Bildschirmfoto 2021-01-28 um 20 26 39" src="https://user-images.githubusercontent.com/1510/106188330-21d56d00-61a7-11eb-8adb-5f5e01165eb6.png">

Author: None are rendering for me right now. I think the service worker might be involved. This reminds me I saw some cache configuration somewhere which we might also need to configure for mp4 files (the videos).

Author: So we'll need to fix that. To temporarily make them show: Develop -> Empty Caches in Safari clears the SW cache and makes the videos appear.

Reviewer 3: 49f3be1 seems to fix it
